import sys
import pandas as pd
from PyQt6 import QtWidgets, uic, QtCore
from google.oauth2.service_account import Credentials
from googleapiclient.discovery import build
import gspread
from dotenv import load_dotenv
import os

load_dotenv()

# ğŸ§  Google Sheets API Servisini oluÅŸtur
def get_service():
    scopes = ['https://www.googleapis.com/auth/spreadsheets']
    creds = Credentials.from_service_account_file(os.getenv("CREDENTIALS_FILE"), scopes=scopes)
    service = build('sheets', 'v4', credentials=creds)
    return service

# ğŸ“¥ Google Sheets'ten veriyi DataFrame olarak Ã§ek
def read_google_sheet():
    #client = gspread.service_account(filename=CREDENTIALS_FILE)
    #working = client.open("Mentor")
    #sayfabir = working.get_worksheet(0)
    #print(sayfabir.get_all_records())


    service = get_service()
    result = service.spreadsheets().values().get(
        spreadsheetId=os.getenv("SPREADSHEET_ID"),
        range=os.getenv("RANGE_NAME"),
    ).execute()

    values = result.get('values', [])
    if not values:
        return pd.DataFrame()
    else:
        return pd.DataFrame(values[1:], columns=values[0])  # ilk satÄ±r baÅŸlÄ±k



# ğŸ§­ PyQt6 ArayÃ¼zÃ¼
class MyApp(QtWidgets.QMainWindow):
    
    def __init__(self,):
        super().__init__()
        uic.loadUi(r".\ui\mentor.ui", self)  # senin UI dosyanÄ±n adÄ±

        #tableWidget ismi senin UI'daki tabloyla aynÄ± olmalÄ±!
        self.load_table_data()


        self.btn_search.clicked.connect(self.search_records)
        self.pushButton.clicked.connect(self.show_all_records)   # TÃ¼m gÃ¶rÃ¼ÅŸmeler butonu

         # ComboBox seÃ§im deÄŸiÅŸtiÄŸinde filtre uygula
        self.comboBox.currentIndexChanged.connect(self.filter_by_combobox)

        self.pushButton_2.clicked.connect(self.close)  # Kapat butonu
    
    def close(self):
        QtWidgets.QApplication.instance().quit() 
    

    def load_table_data(self):
        df = read_google_sheet()
        if df.empty:
            QtWidgets.QMessageBox.warning(self, "UyarÄ±", "Google Sheet'ten veri alÄ±namadÄ±.")
            return

        self.tableWidget.setRowCount(len(df))
        self.tableWidget.setColumnCount(len(df.columns))


        #self.tableWidget.setHorizontalHeaderLabels([str(col) for col in df.columns])
        #bu satir yuzunden df ilk satirini baslik olarak aliyor

        for row in range(len(df)):
            for col in range(len(df.columns)):
                val = str(df.iat[row, col]) if df.iat[row, col] is not None else ""
                self.tableWidget.setItem(row, col, QtWidgets.QTableWidgetItem(val))
        self.df_all = df  # tÃ¼m veriyi sakla
        self.tableWidget.resizeColumnsToContents()

  # ComboBox'u doldur (6. sÃ¼tun = index 5)
        self.comboBox.clear()
        unique_values = sorted(df.iloc[:, 5].dropna().unique())  
        self.comboBox.addItems(unique_values)



    def show_all_records(self):
        self.update_table(self.df_all)
   

    def search_records(self):
        #amac: eski kodda arama yaptigimizda liste guncelleniyordu ve guncellenen listesede sadece arama yaptigimiz isimler oluyordu ondan sonra baska ismi aradigimizda bulamiyorduk
        #guncelleme: isim aramayi her zaman ilk basta olusturdugumuz listeden yaptik.
        #bulunan liste: bulunan isimleri yeni bir listeye ekledik ve o listeyi tabloya aktardik
        #bunu yaparken de update_table adinda yeni bir fonksiyon olusturduk
        #yani kÄ±saca arama yaparken hep orjinal listeyi kullanÄ±yoruz ve bulunanlarÄ± yeni listeye ekleyip tabloya aktarÄ±yoruz
        #bÃ¶ylece arama yaptÄ±ktan sonra baÅŸka bir isim aradÄ±ÄŸÄ±mÄ±zda da bulabiliyoruz
       
        
        #  findData icerisinde arama sonucu bulunan veriler olacak
        findData = []
        
        #kac tane satir var listede. bunun nedenide for dongusu ile her satiri tek tek kontrol edebilmek icin
        row_count = len(self.df_all)  # satÄ±r sayÄ±sÄ±

        # Arama metni
        search_text = self.lineEdit_search.text().lower()  # kÃ¼Ã§Ã¼k harf ile arama

        for row in range(row_count):
            # 2. sÃ¼tundaki deÄŸeri al (0 tabanlÄ± indeksleme)
            val = str(self.df_all.iat[row, 2])  # 2. sÃ¼tun
            # kÃ¼Ã§Ã¼k harf ile karÅŸÄ±laÅŸtÄ±rma
            #arama metni val iÃ§inde geÃ§iyorsa gir
            if search_text in val.lower():  # kÄ±smi eÅŸleÅŸme
                # EÅŸleÅŸen satÄ±rÄ± tablo iÃ§inde seÃ§
                self.tableWidget.selectRow(row)

                # Bulunan satÄ±rÄ± findData listesine ekle
                findData.append([str(self.df_all.iat[row, col]) for col in range(self.df_all.shape[1])])
                print("Bulundu:", val)

        # findData liste olduÄŸu iÃ§in DataFrame'e Ã§evirip update_table ile tabloya aktar
        self.update_table(pd.DataFrame(findData))

    
    def update_table(self,findData):
        # Tabloyu ilk baasta temizle ve yeni veriyi ekle
        #temizleme
        self.tableWidget.setRowCount(0)  # Tabloyu temizle
        #yeni veriyi ekleme
        self.tableWidget.setRowCount(len(findData))  # Yeni satÄ±r sayÄ±sÄ±nÄ± ayarla
        
        for row in range(len(findData)):
            for col in range(len(findData.columns)):
                val = str(findData.iat[row, col]) if findData.iat[row, col] is not None else ""
                self.tableWidget.setItem(row, col, QtWidgets.QTableWidgetItem(val))

        # SÃ¼tun geniÅŸliklerini iÃ§eriÄŸe gÃ¶re ayarla
        self.tableWidget.resizeColumnsToContents()

    def filter_by_combobox(self):
        selected_value = self.comboBox.currentText()

        if not selected_value:
            self.update_table(self.df_all)
            return

        # 5. sÃ¼tuna gÃ¶re filtre uygula
        filtered_df = self.df_all[self.df_all.iloc[:, 5] == selected_value]
        self.update_table(filtered_df)




if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = MyApp()
    window.show()
    sys.exit(app.exec())



# class MyWindow(QtWidgets.QMainWindow):
#     def _init_(self):
#         super(MyWindow, self)._init_()
#         uic.loadUi(r".\ui\admin.ui", self)  # .ui dosyasÄ±nÄ± yÃ¼kler
#         self.pushButton.clicked.connect(self.butonTiklandi)

#     def butonTiklandi(self):
#         print("Butona tÄ±klandÄ±!")

# app = QtWidgets.QApplication(sys.argv)
# window = MyWindow()
# window.show()
# sys.exit(app.exec())  # PyQt6 tarzÄ±